sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/jdbc/JDBCScanBuilder.scala

```diff
@@ -121,6 +123,103 @@ case class JDBCScanBuilder(
     }
   }
 
+  override def isOtherSideCompatibleForJoin(other: SupportsPushDownJoin): Boolean = {
+    other.isInstanceOf[JDBCScanBuilder] &&
+      jdbcOptions.url == other.asInstanceOf[JDBCScanBuilder].jdbcOptions.url
+  };
+
+  // When getJoinedSchema is called, schema shouldn't be pruned yet because pushDownJoin API call
+  // can fail. For this reason, we are temporarily holding pruned schema in new variable that is
+  // later used in pushDownJoin when crafting the SQL query.
+  var aboutToBePrunedSchema: StructType = finalSchema
+
+  override def getJoinedSchema(
+    other: SupportsPushDownJoin,
+    requiredSchema: StructType,
+    otherSideRequiredSchema: StructType): StructType = {
+    aboutToBePrunedSchema = requiredSchema
+    other.asInstanceOf[JDBCScanBuilder].aboutToBePrunedSchema = otherSideRequiredSchema
+
+    var joinedSchema = StructType(Seq())
+
+    (requiredSchema.fields ++ otherSideRequiredSchema.fields)
+      .zipWithIndex
+      .foreach { case (field, idx) =>
+        val newFieldName = JoinOutputAliasIterator.generateColumnAlias
+        joinedSchema =
+          joinedSchema.add(newFieldName, field.dataType, field.nullable, field.metadata)
+      }
+
+    joinedSchema
+  }
+
+  override def pushDownJoin(
+    other: SupportsPushDownJoin,
+    requiredSchema: StructType,
```

Comment:
technically we don't need this parameter as the scan builder can cache it during `getJoinedSchema`?